Distributed actor isolation

DougGregor opened the pull request:
Clean up some aspects of checking distributed actor isolation rules:
* Hopping to a global actor *from* a distributed-actor isolated context does not go through a distributed thunk (global actors can't have one).
* "Unrestricted" declarations are always unrestricted, so we don't need an extra distributed check here (it won't ever occur).
* Actor isolation computation doesn't need a special case for distributed; it marks too much, overriding (e.g.) global actor isolation.
* Model `@_distributedActorIndependent` like `nonisolated`, simplifying away many special cases
* Unify and simplify the checking when performing a cross-actor reference to an entity that is distributed-actor-isolated

DougGregor committed:
Initialize ApplyExpr.ShouldApplyDistributedThunk

DougGregor committed:
Remove unnecessary 'distributed' checks in actor isolation code.

Remove a few places where we are introducing extra distributed-related
actor isolation checking that either isn't necesssary or is incorrect.
Specifically:

* Hopping to a global actor *from* a distributed-actor isolated context
does not go through a distributed thunk (global actors can't have one).
* "Unrestricted" declarations are always unrestricted, so we don't
need an extra distributed check here (it won't ever occur).
* Actor isolation computation doesn't need a special case for
distributed; it marks too much, overriding (e.g.) global actor
isolation.

The primary semantic change visible here is that distributed actors
can now have truly 'nonisolated' members. They aren't required to be
'distributed' because they can't touch state anyway.

ktoso commented on test/Distributed/distributed_actor_isolation.swift:
```diff
@@ -115,6 +114,9 @@ distributed actor DistributedActor_1 {
     try await self.distHelloAsyncThrows()
+
+    // Hops over to the global actor.
+    _ = await DistributedActor_1.staticMainActorFunc()
```
Yay! Glad that just works, we'll need to double check if we reject distributed func getting a global actor -- I'm guessing that's nonsense hmmm

ktoso commented on test/Distributed/distributed_actor_nonisolated.swift:
```diff
@@ -33,7 +31,7 @@ distributed actor DA {
     _ = self.local // expected-error{{distributed actor-isolated property 'local' can only be referenced inside the distributed actor}}
     _ = try await self.dist() // ok, was made implicitly throwing and async
-    _ = self.computedNonisolated // expected-error{{only 'distributed' functions can be called from outside the distributed actor}}
+    _ = self.computedNonisolated // it's okay, only the body of computedNonisolated is wrong
```
Uff, yes absolutely you're right on that

DougGregor committed:
A @_distributedActorIndependent declaration is nonisolated.

The `@__distributedActorIndependent` attribute is effectively the same
as nonisolated, so start treating it that way by making actor-isolation
checking look for it specifically and conclude "nonisolated". Remove
various special cases for this attribute that don't need to exist.

DougGregor committed:
Remove a dead diagnostic

DougGregor committed:
Stop relying on `@_distributedActorIndependent` anywhere.

This is an non-user-visible attribute that is semantically identical to
`nonisolated` except that it allows use in distributed actors. It is
only, and can only, be used for the two fields that distributed actors
have in both the local actor and in the remote proxy, "id" and
"actorTransport". So, special-case the "nonisolated" check for those
fields and stop using `@_distributedActorIndependent`.

DougGregor committed:
Remove `@_distributedActorIndependent` attribute entirely.

All its uses have been subsumed into `nonisolated`.

DougGregor committed:
Remove the unused ActorIsolationRestriction::CrossDistributedActorSelf

DougGregor committed:
Silly fix for mismatched curly braces

DougGregor committed:
Fold a bunch of identical tryMarkImplicitlyAsync calls into one.

ktoso commented on include/swift/AST/Attr.def:
```diff
@@ -658,12 +658,7 @@ CONTEXTUAL_SIMPLE_DECL_ATTR(distributed, DistributedActor,
-  ABIStableToAdd | ABIStableToRemove |
-  APIBreakingToAdd | APIBreakingToRemove,
-  119)
+// 110 is unused
```
Typo, 119 right?

DougGregor replied:
yeah, whoops

yashhchitroda replied:
@DougGregor: can you merge my PR if possible https://github.com/apple/swift/pull/39536

 Sorry for posting this here I have tried reaching all but dint found and solution

ktoso commented on lib/Sema/TypeCheckConcurrency.cpp:
```diff
@@ -385,20 +385,6 @@ ActorIsolationRestriction ActorIsolationRestriction::forDeclaration(
       if (func->isAsyncContext())
         isAccessibleAcrossActors = true;
-
-      // FIXME: move diagnosis out of this function entirely (!)
```
Thanks a lot!

DougGregor committed:
Don't mark calls to async methods of distributed actors implicitly-async

ktoso commented on lib/Sema/TypeCheckConcurrency.h:
```diff
@@ -115,10 +115,6 @@ class ActorIsolationRestriction {
-
-    /// References to declarations that are part of a distributed actor are
-    /// only permitted if they are async.
-    CrossDistributedActorSelf, // FIXME(distributed): remove this case entirely rdar://83713366
```
Great to get rid of it too, yay

DougGregor committed:
Unify 'distributed' checking for accesses to distributed actors.

Unify the logic for checking and marking accesses as
async/throws/uses-distributed-thunk to reduce the overall amount of
code, clarify the isolated/local/potentially-remote cases, and
consistently set these options. There are tiny semantic improvements
here where, e.g., one can asynchronously use non-distributed
functions, properties, and subscripts if we're in a context where we
know the actor is not remote.

DougGregor committed:
Fix incorrect comment in Attr.def

DougGregor commented:
@swift-ci please smoke test

yashhchitroda commented:
@DougGregor @ktoso : can you merge my PR if possible #39536

Sorry for posting this here I have tried reaching all but dint found and solution
